spec/lib/cache_spec.rb

```diff
@@ -121,5 +121,25 @@
         expect(cache.read("my_key")).to eq("bob")
       end
     end
+
+    it "works even when the key expires during fetch" do
+      original_get = Discourse.redis.method(:get)
+
+      calls = 0
+
+      # 1st call returns the value
+      # further calls return nil to simulate the key expiring
+      Discourse
+        .redis
+        .define_singleton_method(:get) do |*args|
+          calls += 1
+          calls == 1 ? Marshal.dump("bob") : nil
+        end
```

Comment:
Looks like we're able to use `.stubs` in other places (e.g. [here](https://github.com/discourse/discourse/blob/e81fc27a0f3c66dd45c1c7b22b4be84491d904e5/spec/services/group_message_spec.rb#L96C35-L96C35))

I would've expected something like this to work:

```
Discourse.redis.stubs(:get).with(key).returns(Marshal.dump("bob")).once
```
(the `.once` asserts it wasn't called a second time)

Reply:
yeah that's what i initially tried but i'm not able to "control" the return values. i wanted a test that similates
the race condition, but maybe ensuring the get is called once is a good enough approximation? 🤔

Reply:
You could also do this to return different values in sequence

```
Discourse.redis.stubs(:get).with(key).returns(Marshal.dump("bob"), nil)
```
but... I think it might fail, since the second value is never actually hit.

IMO checking that `get` is only called once is a very reasonable test for the issue we had.